### ISSUE TITLE
What does embed() return?

### ISSUE THREAD

**Issue by lhqing** [liuhlab/liulab-protein#7]
## Question

What does `embed()` return?

#1 gives `embed(model=..., layer=...) -> np.ndarray` and asserts `p.embed().shape == (33, 960)`
for a 33-residue sequence — per-residue, BOS and EOS stripped. It also notes that a
per-sequence embedding is "a masked mean you write", and that the package ships no pooling
helper.

Open:

- Is per-residue the only v1 return, or does a pooled per-sequence vector `(960,)` need a
  path too? If both, is that a parameter on `embed()` or a second method?
- What `layer=` means and defaults to. `output_hidden_states=True` gives every layer; is the
  default the last hidden state, and is a negative index accepted?
- What `model=` accepts — a checkpoint slug (`esmc_300m`), a HF id (`biohub/ESMC-300M`), or
  a loaded model. And where a loaded model is cached, given that importing `protein` must
  never import torch.
- Whether the array's dtype is pinned, and whether it always comes back on CPU.

Resolve with `/grilling` and `/domain-modeling`; record the vocabulary in `CONTEXT.md`.

**Comment by lhqing** [comment]
embed should not be a protein method, instead, we should have a lightweight class ESM/ESMC/ESMFold to wrap the ESM model family, and the model take protein as input. 

We don't make a general "embed" method on protein. Because embed should be tie to model

**Comment by lhqing** [comment]
## Resolution

`embed()` returns an **`Embedding`** — and it is not a method on `Protein`. Decided with
Hanqing, 2026-09-02.

### `EmbeddingMixin` dissolves

`Protein` gets no embedding method at all: not `embed(model=...)`, and not a named forwarder
either. `embed/` holds a model class.

The rule, which generalises past this ticket: **resident state gets an object; a subprocess
does not.** ESM-C holds 1.33 GB of weights across calls, and a method on `Protein` has nowhere
honest to put them — it either reloads per call or hides a module-level cache keyed on a slug,
a global whose lifetime no caller can see or control. That is the version that fails the moment
someone loops over 500 proteins. mmseqs holds nothing between calls, reaches its database by
path, and already has `external.py` owning the process boundary, so `search()` stays where it
is. The asymmetry between `p.search(db)` and `ESMC().embed(p)` is this rule, not an oversight,
and `CONTEXT.md` should say so.

This narrows #9's direct-support rule rather than contradicting it. That rule is a **ceiling** —
it says where a method may *not* be — and ESM-C does take a sequence. Nothing obliged us to
build up to the ceiling.

Sibling check: genome does both, `STAR(genome, gtf=...).index()` under a thin
`AlignerMixin.build_star_index`. Note which forwarder, though — `build_star_index`, never
`build_index(aligner="star")`. genome already refuses the generic verb. We go one step further
and drop the forwarder too, because unlike an index path, an embedding has no cheap answer to
give without the weights.

### The surface

```python
# embed/embedding.py — numpy only, importable without torch
@dataclass(frozen=True, slots=True)
class Embedding:
    array: np.ndarray      # (L, d_model), float32, CPU, BOS/EOS stripped
    source: str | None     # the accession or chain key it came from
    checkpoint: str        # "300m"
    layer: int             # normalised to a non-negative index

    def __array__(self, dtype=None): ...   # np.asarray(e) is the array
    def __len__(self): ...                 # L
    @property
    def shape(self): ...                   # (L, d_model)
    def mean(self) -> np.ndarray: ...      # (d_model,) — the per-sequence vector

# embed/esm.py — the only module that touches torch, and only inside method bodies
CHECKPOINTS = {"300m": ("biohub/ESMC-300M", 960), "600m": (..., 1152), "6b": (..., 2560)}

class ESMC:
    def __init__(self, checkpoint: str = "300m", *, device: str | None = None,
                 token: str | None = None) -> None
    device: str      # resolved at construction, inspectable
    d_model: int     # from CHECKPOINTS, so 960 is known without loading anything
    n_layers: int    # from the loaded config, so a bad `layer=` names the range
    def embed(self, item: Protein | Chain, *, layer: int = -1) -> Embedding
```

### Every question the ticket asked

**Per-residue or pooled?** Both, and neither is a parameter. `Embedding.mean()`. #1 warned that
a per-sequence vector is "a masked mean you write" — right about the framework, wrong about the
one line. The mask only ever existed for *padded batches*; with BOS and EOS stripped and one
un-padded sequence, `array.mean(axis=0)` is exact. So the pooling helper #1 declined to ship
costs a single method on the returned object.

**`layer=`?** A per-call keyword. Default `-1` = `last_hidden_state`, negatives accepted,
normalised to a non-negative index on the returned `Embedding` so `-1` does not survive into
someone's notes. The default costs no extra memory; a named layer re-runs with
`output_hidden_states=True`. **Unverified and assigned to #13:** how `hidden_states` is indexed
on `EsmcForMaskedLM` — whether element 0 is the embedding layer. #5 verified `last_hidden_state`
and nothing more.

**`model=`?** Gone. It existed because there was nowhere else to put a model. Its three
candidate values resolve as: a **slug** (`"300m"`) is what the constructor takes, an arbitrary
**HF id** is refused, and a **loaded model** is the object you are now holding. The slug wins
over the HF id because a slug validates, names itself in an error, and carries `d_model` in a
module-level table — an arbitrary id is a string that fails somewhere inside `from_pretrained`.
The escape hatch, if ever wanted, is a keyword added later; un-promising an open string is not
possible.

**dtype and device?** Pinned. Always `float32`, always a CPU numpy array, whatever the model ran
on. A caller wanting the tensor on the GPU is doing something v1 does not serve.

### The rest of the shape

- **One class, no base class.** `ESMC` is one architecture with three checkpoints; ESMFold2 is a
  different architecture and is on this map's Out of scope list. An `ESMModel` ABC would be
  extracted from two members when there are two — genome's `Aligner` earns its ABC on shared
  completion records and flag rendering, which `ESMC` and `ESMFold` would not share.
- **Eager construction.** `ESMC()` imports torch, resolves the device, and loads the weights.
  This departs from `Aligner`, which runs nothing at construction — deliberately, because
  `get_index()` constructs a `STAR` just to compute a path, and here the entire point of the
  class is that construction *means* the weights are resident. Cost: `ESMC("600m").d_model`
  would download 2.3 GB to report `1152`, which is why the checkpoint table is module level.
  Doctests that construct one carry `# doctest: +SKIP`.
- **`device=None` auto-resolves** to `cuda` if available else `cpu`, exposed as `.device` so a
  silent CPU fallback is visible — relevant right now, with GPU71FM's GPU unusable (#19).
- **Accepts `Protein` and `Chain`**, the two things in this package that hold a protein
  sequence, and not `str` or a bare `ProteinSequence` — those carry no identity to put in
  `Embedding.source`, and biotite's type is one constructor call from a `Protein`.
- **`Embedding` is a value object, not a bare array**, because a `(33, 960)` array found in a
  notebook an hour later cannot say whether it came from 300M or 600M, or from layer 30 or
  layer 12 — and those arrays are not comparable. `__array__` keeps it numpy-native.
- **One at a time.** Batching stays fog, re-sharpened on the map: #7 removed the reload-per-call
  half of the problem, leaving only the ragged-return question, and `embed()` accepting a list
  later breaks no call that exists.

### Amended

- **#1** — title, opening line, layout tree (`embed/` is `__init__ embedding esm cli`),
  `class Protein(SearchMixin)`, the torch-import rule moved to `embed/esm.py`, a new paragraph
  on `core.py` saying `Protein` has no `embed()`, the mixin section retitled and rewritten with
  both dissolution rules, the pooling sentence corrected, a new **The embedding surface** block,
  and the smoke test now `ESMC().embed(p).shape`.
- **This map** — the Destination line, a new Notes bullet for the resident-state rule, and the
  batching fog patch re-sharpened.
- **#13** — rewritten. It asked for `mixin.py` and `EmbeddingMixin.embed(model=..., layer=...)`.
